### PR TITLE
Faster EventSource attribute lookup

### PR DESCRIPTION
**Pull request by benaadams**
From https://github.com/dotnet/runtime/pull/45121#discussion_r532149600

Contributes to #44598

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2760]: Is `DynamicallyAccessedMembers` still necessary now that we are no longer creating the attribute in this method?

See https://github.com/dotnet/designs/blob/main/accepted/2020/linking-libraries.md#requirements-on-type-values for more information on `DynamicallyAccessedMembers`.

Comment (tommcdon) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2760]: @josalem @sywhang

Comment (benaadams) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs]: Seems ok to remove

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2796]: Note the potential for behavior change here . Previously the first attribute was returned if there were multiple. But now it appears to throw:

https://github.com/dotnet/runtime/blob/8636dc8a76d687169db64033b71ec9f79d9ba4b5/src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs#L505-L515

Looking at the callers of `GetCustomAttributeHelper`, it appears it is only called with `AllowMultiple=false` attributes. But just wanted to call this out.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2767]:
```suggestion
                return member.IsDefined(attributeType, inherit: false);
```

The default `inherit` for `member.IsDefined` is `true`. But the `GetCustomAttributeHelper` method uses `false` below.

https://github.com/dotnet/runtime/blob/8636dc8a76d687169db64033b71ec9f79d9ba4b5/src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs#L475-L478

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (benaadams) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2797]: Added a `Debug.Assert` on type to confirm this contract in code

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2756]:
```suggestion
        // When that is the case, we have to build the custom assemblies on a member by hand.
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2766]:
```suggestion
                // Let the runtime do the work for us, since we can execute code in this context.
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2796]:
```suggestion
                // Let the runtime do the work for us, since we can execute code in this context.
```

**Comment by stephentoub** [comment]
Do you have data on the impact of this?

**Review by noahfalk — APPROVED** [review]
LGTM, thanks Ben!

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (jnyrup) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2763]:
```suggestion
            // System.Diagnostics.Tracing EventSource to be considered valid. This should not matter anywhere but in Microsoft.Diagnostics.Tracing (nuget package).
```

**Comment by benaadams** [comment]
> Do you have data on the impact of this?

Calling into the runtime's `MemberInfo.IsDefined` by default

image

Drops the allocations when the method is called

From:
image

To:
image

As it no longer allocates since https://github.com/dotnet/runtime/pull/44694

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2763]: Then should it be ifdef'd out of corelib?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2791]: Nit:
```suggestion
            Debug.Assert(attributeType == typeof(EventAttribute) || attributeType == typeof(EventSourceAttribute));

```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs
Comment (benaadams) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:2763]: I copied it from `GetCustomAttributeHelper` just below; not sure about what it means in practice?

**Comment by benaadams** [comment]
Rebased to restart CI

**Comment by benaadams** [comment]
Aside: though this has 2 approvals and all tests passed, I can't merge as don't have commit rights

**Review by stephentoub — APPROVED** [review]

**Pull request merged**
